java/ql/src/Telemetry/AutomodelEndpointCharacteristics.qll

```diff
@@ -0,0 +1,325 @@
+
+  string getLocationString(Endpoint e) { result = e.getLocation().toString() }
+
+  predicate isKnownLabel(string label, string humanReadableLabel, EndpointType type) {
+    label = "read-file" and
+    humanReadableLabel = "read file" and
+    type instanceof AutomodelEndpointTypes::TaintedPathSinkType
+    or
+    label = "create-file" and
+    humanReadableLabel = "create file" and
+    type instanceof AutomodelEndpointTypes::TaintedPathSinkType
+    or
+    label = "sql" and
+    humanReadableLabel = "mad modeled sql" and
+    type instanceof AutomodelEndpointTypes::SqlSinkType
+    or
+    label = "open-url" and
+    humanReadableLabel = "open url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "jdbc-url" and
+    humanReadableLabel = "jdbc url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "command-injection" and
+    humanReadableLabel = "command injection" and
+    type instanceof AutomodelEndpointTypes::CommandInjectionSinkType
+  }
+
+  predicate isSink(Endpoint e, string label) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      string input
+    |
+      sinkSpec(e, package, type, subtypes, name, signature, ext, input) and
+      ExternalFlow::sinkModel(package, type, subtypes, name, [signature, ""], ext, input, label, _)
+    )
+  }
+
+  predicate isNeutral(Endpoint e) {
+    exists(string package, string type, string name, string signature |
+      sinkSpec(e, package, type, _, name, signature, _, _) and
+      ExternalFlow::neutralModel(package, type, name, [signature, ""], _)
+    )
+  }
+
+  additional predicate sinkSpec(
+    Endpoint e, string package, string type, boolean subtypes, string name, string signature,
+    string ext, string input
+  ) {
+    package = e.getEnclosingCallable().getDeclaringType().getPackage().toString() and
+    type = e.getEnclosingCallable().getDeclaringType().getName() and
+    subtypes = false and
+    name = e.getEnclosingCallable().getName() and
+    signature = ExternalFlow::paramsString(e.getEnclosingCallable()) and
+    ext = "" and
+    exists(int paramIdx | e.isParameterOf(_, paramIdx) | input = "Argument[" + paramIdx + "]")
+  }
+
+  predicate hasMetadata(Endpoint e, string metadata) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      int input, string provenance, boolean isPublic, boolean isFinal, boolean isStatic,
+      string calleeJavaDoc
+    |
+      hasMetadata(e, package, type, name, signature, input, isFinal, isStatic, isPublic,
+        calleeJavaDoc) and
+      (if isFinal = true or isStatic = true then subtypes = false else subtypes = true) and
+      ext = "" and // see https://github.slack.com/archives/CP9127VUK/p1673979477496069
+      provenance = "ai-generated" and
+      metadata =
+        "{" //
```

Comment:
We have talked about using `$@` placeholders in the extraction queries so that each metadata field gets reported as a separate `relatedLocation` in the SARIF, allowing the downstream code to consume them without JSON-parsing the message string. Shall we start doing that in this PR? I realise the downstream code needs to learn how to consume the new SARIF so that you can test these queries.

Reply:
It's a suggestion I love, it'll remove most (perhaps all?) of the metadata mechanism we currently have. I'll very happily implement it .

One advantage the current mechanism has is that it can return _computed_ values that don't exist like that in source code. In contrast, by using the $@ notation, we can only pass things to the python code that already exist in source code. This, intuitively makes sense to me; but I want to point it out to be sure :)

Reply:
@jhelie, as they're all effectively constants, I'm considering "is public", "provenance", "ext". Does that work for you?

Reply:
OK, I'd like to move this to the next PR to avoid increasing the size of this one :)